docs: document the docs build process (#564)
canonical/charmlibs: CONTRIBUTING.md

Any PR that would trigger a release must also update the library's `CHANGELOG.md` — CI will block the merge otherwise.

Read more: [publishing packages from the monorepo](https://canonical.com/juju/docs/charmlibs/explanation/charmlibs-publishing/).

# Documentation

> The documentation site published at [canonical.com/juju/docs/charmlibs](https://canonical.com/juju/docs/charmlibs) is built with [Sphinx](https://www.sphinx-doc.org/) from the source in the `.docs/` directory. It combines three kinds of content:

Reviewer: Instead of linking to Sphinx, maybe it's more useful (for us - not library maintainers) to say that we use a customised version of https://github.com/canonical/sphinx-stack.

Author: ```suggestion
The documentation site published at [canonical.com/juju/docs/charmlibs](https://canonical.com/juju/docs/charmlibs) uses a custom build process on top of [canonical/sphinx-stack]( https://github.com/canonical/sphinx-stack). It combines three kinds of content:
```

> - **Hand-written diataxis pages** (tutorials, how-to guides, and explanations) authored in `.docs/`, plus per-library pages that live in each library's own `docs/` directory.

Reviewer: ```suggestion
- **Hand-written Diataxis pages** (tutorials, how-to guides, and explanations) authored in `.docs/`, plus per-library pages that live in each library's own `docs/` directory.
```

- **Reference docs** generated automatically from your library's docstrings via Sphinx [autodoc](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html). The module docstring in your package's `__init__.py` becomes the top-level description for that library's reference page.

> - **Interface specification docs** generated from the versioned `README.md` files in each interface's `interface/` directory. These describe the relation interface contract itself (separately from the `charmlibs.interfaces` Python package reference, which is generated from docstrings like any other library).

Reviewer: Re "versioned `README.md` files" - do I remember correctly that we're going to stop supporting versioned interfaces? Should we update the wording here to anticipate that?

Author: Agreed, I tried to remove references to the `v*` directories, but that one slipped though. Thanks for catching it.
```suggestion
- **Interface specification docs** generated from the `README.md` files in each interface's `interface/` directory. These describe the relation interface contract itself (separately from the `charmlibs.interfaces` Python package reference, which is generated from docstrings like any other library).
```

> ## Building the docs

Author: ```suggestion
## Writing library docs

Reference docs are generated from docstrings, so anything you write in your public API's docstrings appears verbatim in the published reference. Keep them informative for library users rather than implementation notes.

For adding tutorials, how-to guides, and explanations specific to your library, see the [how-to guide for adding docs to a library](https://canonical.com/juju/docs/charmlibs/how-to/add-library-docs/).

## Building the docs
```

All docs commands are exposed under the `docs` module of `just`. Run `just docs help` to list them.

```bash
just docs html <package>   # build docs, generating reference docs for <package> only (fast)
just docs html             # build everything, including reference docs for all packages
just docs                  # alias for `just docs html`

> just docs html -           # build the site with no package reference docs (fastest; for editing prose only)

Reviewer: Does prose == Diataxis docs? Since we've called them Diataxis docs above, it might be clearer to call them that here too.

Author: ```suggestion
just docs html -           # build the site with no package reference docs (fastest; for Diataxis docs only)
```

3. **Final combined pass.** A final `sphinx-build` runs with no `package` set. It restores the per-package reference docs saved in step 2, generates the interface specification pages, combines everything with the hand-written pages, and produces the complete HTML site (and `llms.txt`).

The logic for steps 2 and 3 lives in the local Sphinx extensions under `.docs/extensions/`, which are registered in `.docs/conf.py`. In particular:

- `package_docs.py` drives the per-package autodoc passes and saves/restores each library's reference doctree.
- `interface_docs.py` generates the interface specification pages. During the final pass it reads the interface `README.md` files, rewrites relative links to point at the repo on GitHub, and writes the pages under `reference/interfaces/`. (During the per-package passes it only writes a placeholder so the toctree glob doesn't fail.)

> ## Writing library docs

Reviewer: I'd move this section before "Building the docs". That would keep together the main things that library maintainers need to know.

Author: Good call!

Reference docs are generated from docstrings, so anything you write in your public API's docstrings appears verbatim in the published reference. Keep them informative for library users rather than implementation notes.

For adding tutorials, how-to guides, and explanations specific to your library, see the [how-to guide for adding docs to a library](https://canonical.com/juju/docs/charmlibs/how-to/add-library-docs/).

> 

Author: ```suggestion
```

## Working on the docs extensions

The local Sphinx extensions are covered by linting, and have their own tests and static analysis:

```bash
just fast-lint .docs   # run ruff for content under .docs/

> just docs ext-unit     # run unit tests under .docs/tests

Reviewer: ```suggestion
just docs ext-unit     # run unit tests under .docs/tests/
```
